### PR TITLE
util: add class segmented addr

### PR DESCRIPTION
**Pull request by eastonman**
this is used to help gating higher bits of PC (or even any Addr-like reg)

### CODE REVIEW

File: src/main/scala/utility/ClockGatedReg.scala [hunk 1]
```diff
@@ -65,3 +66,79 @@ object GatedRegNextN {
...
+
+class SegmentedAddr(_segments: Seq[Int]) {
+
+  def this(f: Parameters => Seq[Int])(implicit p: Parameters) = this(f(p))
+
+  val segments = _segments // (High, Lower ...)
+  private var addr = UInt(segments.sum.W)
+
+  // [High, Lower ...]
+  private def segment(addr_in: UInt): Seq[UInt] = {
```
Comment (Maxpicca-Li) [inline src/main/scala/utility/ClockGatedReg.scala:78]: Why not uniformly name "addr_in" in Camel?

File: src/main/scala/utility/ClockGatedReg.scala [hunk 2]
```diff
@@ -65,3 +66,79 @@ object GatedRegNextN {
...
+    // Input wire, segmented
+    val segmented = new SegmentedAddr(segments).fromAddr(addr).getAddrSegments()
+
+    val modified = Wire(Vec(segmented.length, Bool()))
+
+    val segmentedNext = segmented zip modified.zipWithIndex map {
+      case (seg, (modified, idx)) => RegEnable(seg, modified && fire)
+        .suggestName(s"${parentName.getOrElse("")}_seg_${idx}_value")
+    }
+    modified zip segmentedNext zip segmented map {
```
Comment (Maxpicca-Li) [inline src/main/scala/utility/ClockGatedReg.scala:131]: It seems that the low bits still need to be compared, so where is the benefit of`SegmentedAddrNext`?

Comment (eastonman) [reply src/main/scala/utility/ClockGatedReg.scala:131]: This is just like GatedRegNext, but for convenience on PC high-low management.

```
  val s1_pc_dup   = s0_pc_dup.zip(io.s0_fire).map {case (s0_pc, s0_fire) => RegEnable(s0_pc, s0_fire)}
  val s2_pc_dup   = s1_pc_dup.zip(io.s1_fire).map {case (s1_pc, s1_fire) => SegmentedAddrNext(s1_pc, pcSegments, s1_fire, Some("s2_pc"))}
  val s3_pc_dup   = s2_pc_dup.zip(io.s2_fire).map {case (s2_pc, s2_fire) => SegmentedAddrNext(s2_pc, s2_fire, Some("s3_pc"))}
```

**Comment by eastonman** [comment]
Q: Why this is needed? Why not use the valid signal of each stage?
A: PC is wide, and higher bits never flip.

Q: Why compare at each register instead of compare once and pass through the pipeline?
A: This has been discussed internally before. PC in BPU can be flushed. There is just no easy way of managing modified bits when the PC is flushed.

File: src/main/scala/utility/ClockGatedReg.scala (hunk 2 above)
Comment (Maxpicca-Li) [reply src/main/scala/utility/ClockGatedReg.scala:133]: I got, thanks.

**Review by Maxpicca-Li — APPROVED** [review]

**Pull request merged**